### ISSUE TITLE
Support for more geometries

### ISSUE THREAD

**Issue by videlec** [flatsurf/flatsurf#302]
I would say that the most general version of flat geometry consists of a bunch of polygons where edges are glued by projective transformations of R^2, that is PGL(3)-structures (this is even more general than what [sage-flatsurf](https://github.com/flatsurf/sage-flatsurf) supports). In a surface endowed with a projective structure we can still talk about lines in the surface (since projective transformations preserve them). There are many interesting subclasses that are obtained by restricting the groups allowed for gluing
- convex PGL(3,R)-structures (without singularities at the vertices) are in bijections with Hitchin representations of surface groups in SL(3,R) ([Choi-Goldman](https://www.jstor.org/stable/2160352))
- when we identify `R^2 = C` one can restrict to `PSL(2,C)` or `PSL(2,Z)`
- `R x SO(2)`-structures (`SimilaritySurface`  in `sage-flatsurf`)
- `SO(2)`-structures (the `ConeSurface` and `RationalConeSurface`)
- `R`-structures and `R_+-structures` (`HalfDilationSurface` and `DilationSurface`)
- `{+/- 1}`-structures (`HalfTranslationSurface`)

**Comment by saraedum** [comment]
> I would say that the most general version of flat geometry consists of a bunch of polygons …

I just want to point out that a "bunch" could be infinite. I don't want to support infinite surfaces in libflatsurf.

As a general comment, I don't think we loose any generality if we assume that all the polygons are triangles. (We need to store the triangulation map in sage-flatsurf to recover the polygons but I think that can be arranged.)

If we keep everything triangulated, then supporting more general types of surfaces comes down to having some optional gluing information that records how half edge `e` is glued to half edge `-e`.

**Comment by saraedum** [comment]
@wphooper wrote at https://github.com/flatsurf/sage-flatsurf/issues/160:

> Perhaps slope coordinates would be useful for libflatsurf?

Currently, we store a vector for each half edge in a triangulated surface. I don't understand how storing slopes makes much of a difference in practice but I think I did not fully understand what you were proposing.

I imagine that the only thing we really need to do here is to drop the requirement that the vector associated to a half edge `v(e)` satisfies that `v(-e)=-v(e)` for the corresponding half edge `-e`.

**Comment by videlec** [comment]
Indeed, a flat structure would associate to each edge a (linear, projective, ...) transformation `m(e)` such that `m(e) v(e) = -v(-e)`. For a translation structure we have `m(e) = identity`. In sage-flatsurf we simplified the possible structures as `m(e)` is always the *unique* similarity `m(e)` mapping `v(e)` to `-v(-e)`.

**Comment by saraedum** [comment]
As discussed today, we probably just want to replace the underlying `OddHalfEdgeMap` with something more general. A structure that does caching of `-` but (virtually?) resolves what `vector(-halfEdge)` when there is a cache miss.

**Comment by videlec** [comment]
For the data structure it might be important to keep in mind that for each kind of surface there is a hidden subgroup of scalar multiplicaion `z -> a z` which is
- the trivial group for `TranslationSurface`
- {+/-1} for `HalfTranslationSurface`
- `R_{> 0}` for `DilationSurface`
- `R \ {0}` for `HalfDilationSurface`
- `{k-th root of unity}` for k-differentials (or `RationalConeSurface` in `sage-flatsurf`)
- `{complex number of modulus one}` for `ConeSurface`
- anything for `SimilaritySurface`

For each half edge, one has to store an element in this group (which is a complex number in general). For translation surfaces we just not store anything since this is always the identity. However, for any other kind of surfaces it is required to have this extra bit of information.
